quinn-proto/src/connection/mod.rs

```diff
@@ -886,18 +885,34 @@ where
     ///
     /// The return value if `Ok` contains the bytes and their offset in the stream.
     pub fn read_unordered(&mut self, id: StreamId) -> Result<Option<(Bytes, u64)>, ReadError> {
-        Ok(self.streams.read_unordered(id)?.map(|result| {
+        let result = self.streams.read_unordered(id)?.map(|result| {
             self.add_read_credits(id, result.max_stream_data, result.max_data);
             (result.buf, result.offset)
-        }))
+        });
+        if self.streams.take_max_streams_dirty(id.dir()) {
```

Comment:
Any issues regarding the early returns of `read` via `?`

I think it might, because there seem occasions where streams are removed on error cases.

Reply:
And consider using `WantTransmit` for those too.

Reply:
Good catch! `ReadError::Reset` indeed can cause stream state to be discarded. `ReadResult` is only available in successful cases, so reusing the `WantTransmit` pattern would require some shuffling around; I'm open to that but I'd like to hold off until after #952 to avoid forcing it to be painfully rebased.